vectordraw/static/js/src/vectordraw.js

```diff
@@ -0,0 +1,607 @@
+            background: null,
+            bounding_box_size: 10,
+            show_navigation: false,
+            show_vector_properties: true,
+            add_vector_label: 'Add Selected Force',
+            vector_properties_label: 'Vector Properties',
+            vectors: [],
+            points: [],
+            expected_result: {},
+            custom_checks: [],
+            unit_vector_ratio: 1
+        };
+        _.defaults(settings, default_settings);
+        var width_scale = settings.width / settings.height,
+            box_size = settings.bounding_box_size;
+        settings.bounding_box = [-box_size*width_scale, box_size, box_size*width_scale, -box_size];
+
+        // Fill in defaults for vectors.
+        var default_vector = {
+            type: 'vector',
+            render: false,
+            length_factor: 1,
+            length_units: '',
+            base_angle: 0,
+            style: {}
+        };
+        var default_vector_style = {
+            pointSize: 1,
+            pointColor: 'red',
+            width: 4,
+            color: "blue",
+            label: null,
+            labelColor: 'black'
+        };
+        settings.vectors.forEach(function(vector) {
+            _.defaults(vector, default_vector);
+            _.defaults(vector.style, default_vector_style);
+        });
+
+        // Fill in defaults for points.
+        var default_point = {
+            fixed: true,  // Default to true for backwards compatibility.
+            render: true,
+            style: {}
+        };
+        var default_point_style = {
+            size: 1,
+            withLabel: false,
+            color: 'pink',
+            showInfoBox: false
+        };
+        settings.points.forEach(function(point) {
+            _.defaults(point, default_point);
+            _.defaults(point.style, default_point_style);
+            point.style.name = point.name;
+            point.style.fixed = point.fixed;
+            point.style.strokeColor = point.style.color;
+            point.style.fillColor = point.style.color;
+            delete point.style.color;
+        });
+
+        return settings;
+    };
+
+    VectorDraw.prototype.render = function() {
+        // Assign the jxgboard element a random unique ID,
+        // because JXG.JSXGraph.initBoard needs it.
+        this.element.find('.jxgboard').prop('id', _.uniqueId('jxgboard'));
+        this.createBoard();
+    };
+
+    VectorDraw.prototype.createBoard = function() {
+        var id = this.element.find('.jxgboard').prop('id'),
+            self = this;
+
+        this.board = JXG.JSXGraph.initBoard(id, {
+            keepaspectratio: true,
+            boundingbox: this.settings.bounding_box,
+            axis: this.settings.axis,
+            showCopyright: false,
+            showNavigation: this.settings.show_navigation
+        });
+
+        function getImageRatio(bg, callback) {
+            $('<img/>').attr('src', bg.src).load(function(){
```

Comment:
I think this code might be causing the issue in Studio with the widget not being correctly initialised on first load.  The load event is kind of tricky and incosistent – see e.g. the caveats in http://api.jquery.com/load-event/.